### PR TITLE
Utf8.fromUtf8(): allow passing length

### PR DESCRIPTION
**Pull request by jpnurmi**

### CODE REVIEW

**Comment by jpnurmi** [comment]
For example: https://github.com/libexpat/libexpat/blob/master/expat/lib/expat.h#L269

**Comment by jpnurmi** [comment]
I had to rebase on top of #60 to be able to run tests.

File: lib/src/utf8.dart
Comment (dcharkes) [inline lib/src/utf8.dart:34]: Utf8 is a variable length encoding. Is this the `length` in encoded bytes? Or the number of characters when decoded?

**Comment by dcharkes** [comment]
@jpnurmi that is a useful addition!

@lrhn Any guidance on whether this should be an optional argument, or whether null-terminated and non-null-terminated strings should use different static methods?

File: lib/src/utf8.dart
Comment (jpnurmi) [reply lib/src/utf8.dart]: It would be the number of one-byte characters, same as `strlen()`. I have added a mention to the docs.

**Comment by jpnurmi** [comment]
- `{int? length}` didn't work with Dart SDK 2.10 from Flutter stable
- `{int length}` wouldn't work with Dart SDK 2.12 from Flutter master

The latest revision uses `{int length = 0}` and checks `length > 0` instead of null to make it work with both older and newer SDKs. I'm also curious about what's the correct way to handle this.

**Comment by dcharkes** [comment]
`package:ffi` is kind of strange, because it works both with and without NNBD at the moment. We will only publish pre-releases opted in to NNBD of the package until we release a stable Dart version with NNBD.

Adding a new static method to differentiate between null-terminated and not will solve the problem for now. @lrhn what do you think from an API design point of view?

Alternatively, you can fork the package and work on your own branch for the time being.

**Comment by jpnurmi** [comment]
I've had similar helpers in a couple of FFI-based projects, so I thought it would be nice to have in `package:ffi`, but I can live with the helpers if it's just causing trouble. :) I think non-null-terminated cases are pretty rare, after all, but it's nice to be able to avoid unnecessary `strlen()` calls in case the length is known.

**Review by lrhn — COMMENTED** [review]
I think an optional argument is fine, it really is two ways to specify the same input to the same function.

A default of `-1` will work both before and after null safety, but that shouldn't be important.
You should *not* be trying to make code which works with both. A library is either null safe or not, and you must pick one.
The `ffi` package has SDK `>= 2.12.0-dev` so it is null safety enabled, and you should be able to use `int? length = null`. If that doesn't work, then null safety (of  the `ffi` package or any other package) is simply not compatible with what you're using it for, and if it works today, it's an accident.

**Comment by jpnurmi** [comment]
> you should be able to use `int? length = null`

`int? length` is fine? Specifying explicitly null complains:

    Don't explicitly initialize variables to null. dart(avoid_init_to_null)

**Comment by lrhn** [comment]
ACK, that's a lint, so you should probably follow it an omit the `= null`. The language itself has no problem with `= null`, but it is the default so you don't have to write it.

File: lib/src/utf8.dart
Comment (lrhn) [inline lib/src/utf8.dart:35]: You can replace ` -- ` by `&mdash;` to get proper em-dashes around the parenthetical phrase.

File: lib/src/utf8.dart
Comment (lrhn) [inline lib/src/utf8.dart:42]: Don't accept both negative values *and* `null` as as signals to use `strlen`. If you do pass an argument, it should be valid. I'd do:
```dart
if (length != null) {
  RangeError.checkNotNegative(length, 'length');
} else {
  length = strlen(strig);
}
```
and then use `length` below (which also avoids the `len` name, we discourage using abbreviations).

File: lib/src/utf8.dart
Comment (jpnurmi) [reply lib/src/utf8.dart:35]: Any preference on "null byte" & "null-terminated" vs. "zero byte" & "zero-terminated"? Looks like I accidentally made it inconsistent because the others are all "null".

File: lib/src/utf8.dart
Comment (jpnurmi) [reply lib/src/utf8.dart]: Heh, you basically wrote the PR for me. Thanks for the help!

File: lib/src/utf8.dart
Comment (lrhn) [reply lib/src/utf8.dart]: The "null terminated" refers to the character (ASCII NUL) and "zero terminated" refers to the byte.
Since value below 128 are both bytes and characters in UTF-8, it's a mostly philosophical question whether you consider the string terminated by the zero byte before decoding it or the NUL character after decoding it.

We do find the length *before* decoding, and it's possible to have an overlong encoding of the NUL character (even if it's invalid), and we won't terminate the string on that, so "zero byte terminated" seems most precise.

(I'd not use `null byte`. Either null/NULL/NUL *character* or zero *byte*.)

**Comment by jpnurmi** [comment]
Let me know if I should push separately the doc changes that are not strictly related to this PR. The Utf8 & Utf16 class descriptions are side by side in the class list so I couldn't leave it like that...

File: lib/src/utf16.dart
Comment (lrhn) [inline lib/src/utf16.dart:11]: Nit: It's not a zero *byte* when it's 16-bit. Maybe a a zero "word" or "code unit". Maybe just "zero terminated", then it's the zero value of the appropriate type.

File: lib/src/utf16.dart
Comment (lrhn) [inline lib/src/utf16.dart:21]: Should probably say UTF-16 here.
(You don't have to fix all existing typos, I'm just mentioning it so someone, like @dcharkes, can fix it later). :wink:

File: lib/src/utf16.dart
Comment (lrhn) [inline lib/src/utf16.dart:17]: Ditto: zero terminated.

File: lib/src/utf16.dart
Comment (lrhn) [inline lib/src/utf16.dart:19]: ... contains a zero code unit  ....

Comment (lrhn) [reply lib/src/utf16.dart:19]: It actually won't be truncated. It allocates and copies all the code units over.
If that result is interpreted as zero-terminated, then that result will be truncated, but if you know the length, you should be able to get it all back.
(Not your text either.)

File: lib/src/utf8.dart
Comment (lrhn) [inline lib/src/utf8.dart:28]: Idea: Maybe use `indexOf(0)` below instead of `indexWhere(closure)`. Should have less overhead.

File: test/utf8_test.dart
Comment (lrhn) [reply test/utf8_test.dart:91]: Excellent :)

**Review by lrhn — COMMENTED** [review]
I think this look fine, apart from referring to UTF-16 code units as "zero byte"s.
I've made some extra notes, but that's just drive-by comments on things you haven't touched. I'd let @dcharkes ponder those.
Also, it's his package, so I'll let him decide how to structure CLs (I'm a fan of monolithic CLs that fix everything in one go. Other people tend to disagree). 😁

**Comment by jpnurmi** [comment]
I opened a separate less intrusive PR to update the docs #63. I'll rebase and cleanup this afterwards if @dcharkes agrees with the changes.

**Review by dcharkes — APPROVED** [review]

**Comment by jpnurmi** [comment]
Thanks!

P.S. There's still one `strlen()` improvement idea buried in the comments:

> lrhn (Lasse R.H. Nielsen) 15 hours ago Member
> Idea: Maybe use indexOf(0) below instead of indexWhere(closure). Should have less overhead.

**Comment by dcharkes** [comment]
@jpnurmi, 1 thing is missing for publishing a new version of the package. Please make a PR that bumps the version number to `0.2.0-nullsafety.1` and add an entry to `CHANGELOG.md`. Then I can release another pre-release (only usable in Dart 1.12).

Alternatively, you can just depend on a git hash instead (or your own branch) for the time being until Dart 1.12 is released to stable.

**Comment by dcharkes** [comment]
> Thanks!
> 
> P.S. There's still one `strlen()` improvement idea buried in the comments:
> 
> > lrhn (Lasse R.H. Nielsen) 15 hours ago Member
> > Idea: Maybe use indexOf(0) below instead of indexWhere(closure). Should have less overhead.

Do you want to address this before I merge #64?

**Comment by jpnurmi** [comment]
I don't mind :) => #65

**Pull request merged**
